src/v/v8_engine/script_dispatcher.h

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "model/metadata.h"
+
+#include <seastar/core/condition-variable.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/shared_ptr.hh>
+
+#include <absl/container/node_hash_map.h>
+#include <absl/container/node_hash_set.h>
+#include <v8_engine/script.h>
+
+namespace v8_engine {
+
+struct script_compile_params {
+    script_compile_params(ss::sstring function_name, iobuf code)
+      : function_name(std::move(function_name))
+      , code(std::move(code)) {}
+
+    ss::sstring function_name;
+    iobuf code;
+    size_t max_heap_size_bytes{10000};
+    size_t run_timeout_ms{200};
+};
+
+template<typename Executor>
+class script_dispatcher {
+public:
+    explicit script_dispatcher(Executor& executor)
+      : _executor(executor) {}
+
+    // Insert only put compilation params to map.
+    bool insert(
+      const model::topic_namespace& topic,
+      ss::sstring function_name,
+      iobuf code) {
+        return _compilation_params
+          .emplace(
+            topic,
+            script_compile_params(std::move(function_name), std::move(code)))
+          .second;
+    }
+
+    bool remove(const model::topic_namespace& topic) {
+        bool contains_script_flag = _compilation_params.erase(topic) == 1;
+        if (!contains_script_flag) {
+            return false;
+        }
+
+        auto runtime = _runtimes.find(topic);
+        if (runtime == _runtimes.end()) {
+            return true;
+        }
+
+        if (runtime->second.get()->function.get() == nullptr) {
+            runtime->second->is_deleted = true;
+            runtime->second->cv.broadcast();
+        }
+
+        _runtimes.erase(runtime);
+
+        return true;
+    }
+
+    ss::future<ss::lw_shared_ptr<script>> get(model::topic_namespace topic) {
+        auto params_it = _compilation_params.find(topic);
+        if (params_it == _compilation_params.end()) {
+            co_return ss::lw_shared_ptr<script>(nullptr);
+        }
+
+        auto runtime_it = _runtimes.find(topic);
+        if (runtime_it == _runtimes.end()) {
+            auto value = ss::make_lw_shared(runtimes_value());
+            _runtimes.emplace(value);
+
+            auto runtime = ss::make_lw_shared(script(
+              params_it->second.max_heap_size_bytes,
+              params_it->second.run_timeout_ms));
+
+            try {
+                co_await runtime->init(
```

Comment:
As it stands delete doesn't cancel an ongoing compile, maybe thats a desired feature?